File: Huffman.java
package DAA;
import java.util.*;

class HuffmanNode {
    int data;
    char c;
    HuffmanNode left;
    HuffmanNode right;
}

class MyComparator implements Comparator<HuffmanNode> {
    public int compare(HuffmanNode x, HuffmanNode y) {
        return x.data - y.data;
    }
}

public class Huffman {
    public static void printHuffmanCodes(HuffmanNode root, String s) {
        if (root.left == null && root.right == null ) {
            System.out.println(root.c + ":" + s);
            return;
        }
        printHuffmanCodes(root.left, s + "0");
        printHuffmanCodes(root.right, s + "1");
    }

    public static void buildHuffmanTree(char[] charArray, int[] charFreq) {
        PriorityQueue<HuffmanNode> priorityQueue = new PriorityQueue<>(charArray.length, new MyComparator());

        for (int i = 0; i < charArray.length; i++) {
            HuffmanNode node = new HuffmanNode();
            node.c = charArray[i];
            node.data = charFreq[i];
            node.left = null;
            node.right = null;
            priorityQueue.add(node);
        }

        while (priorityQueue.size() > 1) {
            HuffmanNode x = priorityQueue.poll();
            HuffmanNode y = priorityQueue.poll();
            HuffmanNode newNode = new HuffmanNode();
            newNode.data = x.data + y.data;
            newNode.c = '-';
            newNode.left = x;
            newNode.right = y;
            priorityQueue.add(newNode);
        }
        printHuffmanCodes(priorityQueue.poll(), "");
    }

    public static void main(String[] args) {
        char[] charArray = { 'a', 'b', 'c', 'd', 'e', 'f' };
        int[] charFreq = { 5, 9, 12, 13, 16, 45 };

        buildHuffmanTree(charArray, charFreq);
    }
}

File: huffman.py
import heapq

class MinHeapNode:
    def _init_(self, data, freq):
        self.data = data
        self.freq = freq
        self.left = None
        self.right = None

def print_codes(root, string):
    if root is None:
        return
    if root.data != '$':
        print(root.data + ": " + string)
    print_codes(root.left, string + "0")
    print_codes(root.right, string + "1")

def huffman_code(data, freq):
    min_heap = []

    for i in range(len(data)):
        node = MinHeapNode(data[i], freq[i])
        heapq.heappush(min_heap, (node.freq, id(node), node))

    while len(min_heap) > 1:
        freq1, _, left = heapq.heappop(min_heap)
        freq2, _, right = heapq.heappop(min_heap)
        temp = MinHeapNode('$', freq1 + freq2)
        temp.left = left
        temp.right = right
        heapq.heappush(min_heap, (temp.freq, id(temp), temp))

    root = min_heap[0][2]
    print("Huffman Codes:")
    print_codes(root, "")

if __name__ == "__main__":
    n = int(input("Enter total number of  character to insert: "))
    m = int(input("Enter total number of  frequency to insert: "))
    data = []
    freq = []

    for i in range(n):
        char = input("Enter the character: ")
        data.append(char)
    for i in range(n):
        f = int(input("Enter the frequency: "))
        freq.append(f)

    huffman_code(data, freq)